cmd/fs-v1-background-append.go

```diff
@@ -0,0 +1,212 @@
+// can not append and have to wait till part-1 is uploaded. Hence we return
+// this error. Currently this error is not used in the caller.
+var errPartsMissing = errors.New("required parts missing")
+
+// Error sent when appendParts go-routine has waited long enough and timedout.
+var errAppendPartsTimeout = errors.New("appendParts goroutine timeout")
+
+// Timeout value for the appendParts go-routine.
+var appendPartsTimeout = 24 * 60 * 60 * time.Second
+
+// Holds a map of uploadID->appendParts go-routine
+type backgroundAppend struct {
+	infoMap map[string]bgAppendPartsInfo
+	sync.Mutex
+}
+
+// Input to the appendParts go-routine
+type bgAppendPartsInput struct {
+	meta  fsMetaV1   // list of parts that need to be appended
+	errCh chan error // error sent by appendParts go-routine
+}
+
+// Identifies an appendParts go-routine.
+type bgAppendPartsInfo struct {
+	inputCh   chan bgAppendPartsInput
+	timeoutCh chan struct{} // closed by appendParts go-routine when it timesout
+	endCh     chan struct{} // closed after complete/abort of upload to end the appendParts go-routine
+}
+
+// Called after a part is uploaded so that it can be appended in the background.
+func (b *backgroundAppend) append(disk StorageAPI, bucket, object, uploadID string, meta fsMetaV1) {
+	b.Lock()
+	info, ok := b.infoMap[uploadID]
+	if !ok {
+		// Corresponding appendParts go-routine was not found, create a new one. Would happen when the first
+		// part of a multipart upload is uploaded.
+		inputCh := make(chan bgAppendPartsInput)
+		timeoutCh := make(chan struct{})
+		endCh := make(chan struct{})
+
+		info = bgAppendPartsInfo{inputCh, timeoutCh, endCh}
+		b.infoMap[uploadID] = info
+
+		go b.appendParts(disk, bucket, object, uploadID, info)
+	}
+	b.Unlock()
+	go func() {
+		errCh := make(chan error)
+		// send input in a goroutine as send on the inputCh can block if appendParts go-routine
+		// is busy appending a part.
+		select {
+		case <-info.timeoutCh:
+			// This is to handle a rare race condition where we found info in b.infoMap
+			// but soon after that appendParts go-routine timed out.
+		case info.inputCh <- bgAppendPartsInput{meta, errCh}:
+			// Receive the error so that the appendParts go-routine does not block on send.
+			// But the error received is ignored as fs.PutObjectPart() would have already
+			// returned success to the client.
+			<-errCh
+		}
+	}()
+}
+
+// Called on complete-multipart-upload. Returns nil if app the required parts have been appended.
+func (b *backgroundAppend) complete(disk StorageAPI, bucket, object, uploadID string, meta fsMetaV1) error {
+	b.Lock()
+	info, ok := b.infoMap[uploadID]
+	b.Unlock()
+	if !ok {
+		return errPartsMissing
+	}
+	errCh := make(chan error)
+	select {
+	case <-info.timeoutCh:
+		// This is to handle a rare race condition where we found info in b.infoMap
+		// but soon after that appendParts go-routine timedouted out.
+		return errAppendPartsTimeout
```

Comment:
the necessary cleanup happens when the timeout happens inside the go-routine:
```
                case <-time.After(appendPartsTimeout):                                                                                                                          
                        // Timeout the goroutine to garbage collect its resources. This would happen if the client initates                                                     
                        // a multipart upload and does not complete/abort it.                                                                                                   
                        b.Lock()                                                                                                                                                
                        delete(b.infoMap, uploadID)                                                                                                                             
                        b.Unlock()                                                                                                                                              
                        // Delete the temporary append file as well.                                                                                                            
                        appendFilePath := getFSAppendDataPath(uploadID)                                                                                                         
                        disk.DeleteFile(bucket, appendFilePath)                                                                                                                 
                                                                                                                                                                                
                        close(info.timeoutCh)
```